### PR TITLE
IIR Refactoring

### PR DESCRIPTION
**Pull request by twicki**
# Technical Description

We removed the dependency on the SIR and moved all the data used into metadata when appropriate and the rest to derived information where it was needed

### CODE REVIEW

**Comment by twicki** [comment]
launch pipeline

**Comment by twicki** [comment]
launch pipeline

File: src/dawn/IIR/MetaInformation.h [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+  /// variables
+  VariableVersions variableVersions_;
+
+  /// Stencil description statements. These are built from the StencilDescAst of the sir::Stencil
+  std::vector<std::shared_ptr<Statement>> stencilDescStatements_;
+  std::unordered_map<std::shared_ptr<StencilCallDeclStmt>, int> StencilCallToStencilIDMap_;
+  std::unordered_map<int, std::shared_ptr<StencilCallDeclStmt>> IDToStencilCallMap_;
+
+  /// StageID to name Map. Filled by the `PassSetStageName`.
+  std::unordered_map<int, std::string> StageIDToNameMap_;
```
Comment (cosunae) [inline src/dawn/IIR/MetaInformation.h]: it looks to me like this does not belong here. It is redundant, the information of the name of the stage is already in the Stage

File: src/dawn/IIR/MetaInformation.h [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+                     std::shared_ptr<StencilFunctionInstantiation>>
+      ExprToStencilFunctionInstantiationMap_;
+
+  /// lookup table containing all the stencil function candidates, whose arguments are not yet bound
+  std::unordered_map<std::shared_ptr<StencilFunctionInstantiation>,
+                     StencilFunctionInstantiationCandidate>
+      stencilFunInstantiationCandidate_;
+
+  /// BoundaryConditionCall to Extent Map. Filled my `PassSetBoundaryCondition`
+  std::unordered_map<std::shared_ptr<BoundaryConditionDeclStmt>, Extents>
```
Comment (cosunae) [inline src/dawn/IIR/MetaInformation.h]: I think that iir::Stencil should contain a iir::BoundaryCondition, and that one contain this map

File: src/dawn/IIR/MetaInformation.h [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+
+  /// Field Name to BoundaryConditionDeclStmt
+  std::unordered_map<std::string, std::shared_ptr<BoundaryConditionDeclStmt>>
+      FieldnameToBoundaryConditionMap_;
+
+  /// Set of all the IDs that are locally cached
+  std::set<int> CachedVariableSet_;
+
+  /// Map of Field ID's to their respecive legal dimensions for offsets if specified in the code
+  std::unordered_map<int, Array3i> fieldIDToInitializedDimensionsMap_;
```
Comment (cosunae) [inline src/dawn/IIR/MetaInformation.h]: the naming here I think is confusing

File: src/dawn/Compiler/DawnCompiler.cpp [hunk 4]
```diff
@@ -186,8 +186,10 @@ std::unique_ptr<OptimizerContext> DawnCompiler::runOptimizer(std::shared_ptr<SIR
     std::shared_ptr<iir::StencilInstantiation> instantiation = stencil.second;
     DAWN_LOG(INFO) << "Starting Optimization and Analysis passes for `" << instantiation->getName()
                    << "` ...";
+//    instantiation->dump();
```
Comment (cosunae) [inline src/dawn/Compiler/DawnCompiler.cpp]: remove

File: src/dawn/IIR/MetaInformation.cpp [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+  for(auto pair : origin.StmtToAccessIDMap_) {
+    StmtToAccessIDMap_.emplace(pair.first->clone(), pair.second);
+  }
+  LiteralAccessIDToNameMap_ = origin.LiteralAccessIDToNameMap_;
+  FieldAccessIDSet_ = origin.FieldAccessIDSet_;
+  apiFieldIDs_ = origin.apiFieldIDs_;
+  TemporaryFieldAccessIDSet_ = origin.TemporaryFieldAccessIDSet_;
+  AllocatedFieldAccessIDSet_ = origin.AllocatedFieldAccessIDSet_;
+  GlobalVariableAccessIDSet_ = origin.GlobalVariableAccessIDSet_;
+  for(auto id : origin.variableVersions_.getVersionIDs()) {
```
Comment (cosunae) [inline src/dawn/IIR/MetaInformation.cpp]:
```suggestion
  variableVersions_.insert(origin.variableVersions_.begin(), origin.variableVersions_.end());
```

File: src/dawn/IIR/MetaInformation.h [hunk 6]
```diff
@@ -0,0 +1,493 @@
...
+#include <set>
+#include <string>
+#include <unordered_map>
+
+namespace dawn {
+namespace iir {
+
+/// @brief Specific instantiation of a stencil
+/// @ingroup optimizer
+class StencilMetaInformation : NonCopyable {
```
[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/IIR/MetaInformation.h]: fixed

File: src/dawn/IIR/MetaInformation.h (hunk 1 above)
Comment (twicki) [reply src/dawn/IIR/MetaInformation.h]: fixed

File: src/dawn/IIR/MetaInformation.h (hunk 3 above)
Comment (twicki) [reply src/dawn/IIR/MetaInformation.h]: why, these are the dimensions that the user initialized? what would be your suggetstion?

File: src/dawn/Compiler/DawnCompiler.cpp (hunk 4 above)
Comment (twicki) [reply src/dawn/Compiler/DawnCompiler.cpp]: done

File: src/dawn/IIR/MetaInformation.h (hunk 2 above)
Comment (twicki) [reply src/dawn/IIR/MetaInformation.h]: why should this be on the level of the iir::Stencil? I think this should be on the level of each user-stencil has this associated with it. and this is the level of the steniclInstantiation

**Comment by twicki** [comment]
launch pipeline

**Comment by twicki** [comment]
launch pipeline

**Comment by twicki** [comment]
launch pipeline

**Comment by twicki** [comment]
launch pipeline

File: src/dawn/Optimizer/PassSetNonTempCaches.cpp [hunk 7]
```diff
@@ -133,8 +133,8 @@ class GlobalFieldCacher {
       originalNameToCache_.emplace_back(
           NameToImprovementMetric{instantiation_->getOriginalNameFromAccessID(oldID), cache,
                                   accessIDToDataLocality_.find(oldID)->second});
-      instantiation_->insertCachedVariable(oldID);
-      instantiation_->insertCachedVariable(newID);
+//      instantiation_->insertCachedVariable(oldID);
```
[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/Optimizer/PassSetNonTempCaches.cpp]: this was a deprecated structure and is now removed

File: src/dawn/IIR/StencilMetaInformation.cpp [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    candidate.callerStencilFunction_ =
+        std::make_shared<StencilFunctionInstantiation>(pair.second.callerStencilFunction_->clone());
+    stencilFunInstantiationCandidate_.emplace(
+        std::make_shared<StencilFunctionInstantiation>(pair.first->clone()), candidate);
+  }
+  for(const auto& pair : origin.FieldnameToBoundaryConditionMap_) {
+    FieldnameToBoundaryConditionMap_.emplace(
+        pair.first, std::make_shared<BoundaryConditionDeclStmt>(*(pair.second)));
+  }
+  //  CachedVariableSet_ = origin.CachedVariableSet_;
```
[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/IIR/StencilMetaInformation.cpp]: done

**Comment by twicki** [comment]
launch pipeline

**Pull request merged**